### PR TITLE
Use std::vector instead of dynamic C-Array

### LINKED ISSUE #6740: MSVC compiler error mg_decoration.cpp(189): error C2131: size expression is not constant
Localized error message:
```
2>\Minetest\src\mapgen\mg_decoration.cpp(189): error C2131: Ausdruck wurde nicht zu einer Konstanten ausgewertet.
2>\Minetest\src\mapgen\mg_decoration.cpp(189): note: Der Fehler wurde durch einen Lesezugriff auf eine Variable außerhalb ihrer Lebensdauer verursacht.
2>\Minetest\src\mapgen\mg_decoration.cpp(189): note: siehe Verwendung von "size"
2>\Minetest\src\mapgen\mg_decoration.cpp(190): error C2131: Ausdruck wurde nicht zu einer Konstanten ausgewertet.
2>\Minetest\src\mapgen\mg_decoration.cpp(190): note: Der Fehler wurde durch einen Lesezugriff auf eine Variable außerhalb ihrer Lebensdauer verursacht.
2>\Minetest\src\mapgen\mg_decoration.cpp(190): note: siehe Verwendung von "size"
```

### PR DESCRIPTION
**Pull request by adrido**
Fixes #6740

I'm wondering, if it would possible/better to use a dynamic vector and `floors.size()` instead of `num_floors` and `num_ceilings` and a fixed vector.
But I does not want to break existing code and this PR is only meant to fix the failing build not for doing improvements.

### CODE REVIEW

File: src/mapgen/mapgen.h [hunk 1]
```diff
@@ -194,7 +194,8 @@ class Mapgen {
 	s16 findLiquidSurface(v2s16 p2d, s16 ymin, s16 ymax);
 	void updateHeightmap(v3s16 nmin, v3s16 nmax);
 	void getSurfaces(v2s16 p2d, s16 ymin, s16 ymax,
-		s16 *floors, s16 *ceilings, u16 *num_floors, u16 *num_ceilings);
+		std::vector<s16> &floors, std::vector<s16> &ceilings,
+		u16 * num_floors, u16 * num_ceilings);
```
Comment (SmallJoker) [inline src/mapgen/mapgen.h]: Pointer misplaced. `u16 *num_floors`, same for `num_ceilings`

**Comment by SmallJoker** [comment]
C-style arrays were used for speed reasons, as the maximal size of the data amount is known.
You could still allocate the memory directly using `s16 *floors = new s16[size]` + `delete[] floors` or `std::array<s16, size> floors`.

**Comment by paramat** [comment]
Yes that's my code, i was unsure whether to use a vector instead of an array.

**Comment by adrido** [comment]
>I suppose the C-style arrays were used for speed reasons, as the maximal size of the data amount is known. (ask @paramat ?)
You could still allocate the memory directly using s16 *floors = new s16[size] + delete[] floors or std::array<s16, size> floors.

Thanks, I know. Its just my suggestion to fix #6740 . I don't know whether this function is time critical or not. A vector is just a more secure variant (and not necessary slower at all). Feel free to fix #6740 with another way. Every variant would fine for me as long as it fixes #6740. The merge process might be faster if a coredev is doing this.

**Comment by paramat** [comment]
I've seen vector used in many places in mapgen so i doubt it's slower, or significantly slower. I probably should have used vector in my PR anyway.

**Comment by paramat** [comment]
Works fine :+1: 
I'm no expert concerning the relative advantages of one against the other.

**Comment by adrido** [comment]
Thanks. A probably interesting article of C-Array vs std::array vs std::vector might be this by Bjarne Stroustrup: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#slcon1-prefer-using-stl-array-or-vector-instead-of-a-c-array

**Comment by paramat** [comment]
Good, vector is fine for me.

**Comment by adrido** [comment]
@paramat I added some improvements that got possible due to the usage of vectors. Could you please test it again?

**Comment by paramat** [comment]
Interesting, i'll rereview to understand and then retest. Removed my +1 for now.

**Comment by paramat** [comment]
Code is fine, will test.

**Comment by paramat** [comment]
Ceiling decorations are not working, i suspect logic, maybe due to the removal of 'num_floors/ceilings'.

File: src/mapgen/mapgen.cpp [hunk 2]
```diff
@@ -318,19 +316,14 @@ void Mapgen::getSurfaces(v2s16 p2d, s16 ymin, s16 ymax,
 		is_walkable = ndef->get(mn).walkable;
 
 		if (is_walkable && !walkable_above) {
-			floors[floor_i] = y;
-			floor_i++;
+			floors.push_back(y);
 		} else if (!is_walkable && walkable_above) {
-			ceilings[ceiling_i] = y + 1;
-			ceiling_i++;
+			ceilings.push_back(y);
```
Comment (paramat) [inline src/mapgen/mapgen.cpp]: `ceilings.push_back(y + 1);`

**Comment by adrido** [comment]
Thanks for reviewing. Done.
I just want to note: 
1. ) std::array will not work here too, because size is still not const (at compile time). GCC or Clang may support it, but its not official standard of C++

2. ) 
Stroustrup: 
   > People sometimes worry about the cost of std::vector growing incrementally. I used to worry about that and used reserve() to optimize the growth. After measuring my code and repeatedly having trouble finding the performance benefits of reserve() in real programs, I stopped using it except where it is needed to avoid iterator invalidation (a rare case in my code). Again: measure before you optimize. 
 
   http://www.stroustrup.com/bs_faq2.html#slow-containers

**Comment by paramat** [comment]
Will retest then merge.

**Comment by paramat** [comment]
Tested and fine :+1:

**Pull request merged**